data-plane/dispatcher/src/main/java/dev/knative/eventing/kafka/broker/dispatcher/strategy/UnorderedConsumerRecordOffsetStrategy.java

```diff
@@ -57,64 +66,101 @@ public UnorderedConsumerRecordOffsetStrategy(final KafkaConsumer<?, ?> consumer,
   }
 
   /**
    * {@inheritDoc}
    */
   @Override
   public void failedToSendToDLQ(final KafkaConsumerRecord<?, ?> record, final Throwable ex) {
-    // do not commit
+    mutateStateAndCheckAck(new TopicPartition(record.topic(), record.partition()), record.offset());
   }
 
   /**
    * {@inheritDoc}
    */
   @Override
   public void recordDiscarded(final KafkaConsumerRecord<?, ?> record) {
-    successfullySentToSubscriber(record);
+    mutateStateAndCheckAck(new TopicPartition(record.topic(), record.partition()), record.offset());
   }
 
-  private Future<Map<TopicPartition, OffsetAndMetadata>> commit(
-    final KafkaConsumerRecord<?, ?> record) {
-    logger.debug("committing record {}", record);
-    return consumer.commit(Map.of(
-      new TopicPartition(record.topic(), record.partition()),
-      new OffsetAndMetadata(record.offset() + 1, ""))
-    );
+  /**
+   * @return null if it shouldn't ack, otherwise the offset to ack - 1.
+   */
+  private Long mutateStateAndCheckAck(TopicPartition topicPartition, long offset) {
+    long lastAckedOffset = this.lastAckedMap.get(topicPartition); // This is always non null
+    SortedSet<Long> toAckSet = this.pendingAcksMap.computeIfAbsent(topicPartition, v -> new TreeSet<>());
+    toAckSet.add(offset);
+
+    return (toAckSet.first() == lastAckedOffset + 1 && toAckSet.last() - toAckSet.first() == toAckSet.size() - 1) ?
+      toAckSet.last() : null;
   }
+
+  private void commit(final KafkaConsumerRecord<?, ?> record) {
+    TopicPartition key = new TopicPartition(record.topic(), record.partition());
+    Long shouldAck = mutateStateAndCheckAck(key, record.offset());
+    if (shouldAck != null) {
+      // Reset the state
+      long lastAckedBeforeThisOne = this.lastAckedMap.put(key, shouldAck);
+      SortedSet<Long> messagesImGoingToAck = this.pendingAcksMap.remove(key);
+
+      // Execute the actual commit
+      consumer.commit(Map.of(
+        key,
+        new OffsetAndMetadata(shouldAck + 1, ""))
+      ).onSuccess(ignored -> {
+        eventsSentCounter.increment(messagesImGoingToAck.size());
+        logger.debug(
+          "committed {} {} {}",
+          keyValue("topic", record.topic()),
+          keyValue("partition", record.partition()),
+          keyValue("offset", shouldAck)
+        );
+      })
+        .onFailure(cause -> {
+          // If the commit failed, there are 2 situations:
+          // * Nobody tried to commit again, so let's just restore the state
+          // * Somebody committed with an offset greater than this one, so we just discard the error
+          if (!(this.lastAckedMap.get(key) > shouldAck)) {
+            this.lastAckedMap.put(key, lastAckedBeforeThisOne);
+            this.pendingAcksMap.compute(key, (k, actual) -> {
+              if (actual != null) {
+                actual.addAll(messagesImGoingToAck);
+                return actual;
+              }
+              return messagesImGoingToAck;
+            });
+          }
```

Comment:
This needs some testing.

Reply:
It's already tested, check out `shouldCommitInAMixedOrderWithInducedFailure` and `shouldCommitAfterSendingEventsOrderedOnTheSamePartitionWithInducedFailure`

Reply:
If I comment this code, tests are green. :confused: